In [ ]:
from google.colab import drive
drive.mount('/content/drive')

In [ ]:
!pip install torch==1.6.0+cu101 torchvision==0.7.0+cu101 -f https://download.pytorch.org/whl/torch_stable.html

In [ ]:
!pip install console-progressbar

In [ ]:
!pip install fastai==1.0.61

In [ ]:
%reload_ext autoreload
%autoreload 2
%matplotlib inline
from fastai.vision import *
from fastai.metrics import error_rate
from fastai.vision.data import ImageDataBunch
from fastai import *
import cv2 as cv
import numpy as np
import pandas as pd
import scipy.io as sio
import zipfile


#Reference taken for the code 
# dhamvi01(2019),FastAI-Image-Classification
# https://github.com/dhamvi01/FastAI-Image-Classification/blob/master/fastai.ipynb

# Data prprocessing

# Prepare data batches

In [ ]:

data = ImageDataBunch.from_folder('/content/drive/MyDrive/cars_classification/cars_classification_dataset','train','valid',ds_tfms=get_transforms(do_flip=False, flip_vert=True, max_rotate=5.0, max_zoom=1.1, max_lighting=0.2, max_warp=0.2, p_affine=0.75, p_lighting=0.75),size=224,bs=32).normalize(imagenet_stats)

In [ ]:
# show the batches of image tha created by the model itself...
data.show_batch()

In [ ]:
# class names and number of classes
print(data.classes)
len(data.classes),data.c

['1', '10', '100', '101', '102', '103', '104', '105', '106', '107', '108', '109', '11', '110', '111', '112', '113', '114', '115', '116', '117', '118', '119', '12', '120', '121', '122', '123', '124', '125', '126', '127', '128', '129', '13', '130', '131', '132', '133', '134', '135', '136', '137', '138', '139', '14', '140', '141', '142', '143', '144', '145', '146', '147', '148', '149', '15', '150', '151', '152', '153', '154', '155', '156', '157', '158', '159', '16', '160', '161', '162', '163', '164', '165', '166', '167', '168', '169', '17', '170', '171', '172', '173', '174', '175', '176', '177', '178', '179', '18', '180', '181', '182', '183', '184', '185', '186', '187', '188', '189', '19', '190', '191', '192', '193', '194', '195', '196', '2', '20', '21', '22', '23', '24', '25', '26', '27', '28', '29', '3', '30', '31', '32', '33', '34', '35', '36', '37', '38', '39', '4', '40', '41', '42', '43', '44', '45', '46', '47', '48', '49', '5', '50', '51', '52', '53', '54', '55', '56', '57', '58', '

(196, 196)

# Training: ResNet 152

# For the training, using ResNet152 and batch size of 32.

In [ ]:
epochs = 40
learn = cnn_learner(data, models.resnet152, metrics=accuracy)
learn.fit(epochs)


data = learn.recorder.plot_losses()
data_acc = learn.recorder.plot_metrics()

steps_per_epoch = int(len(data[1])/epochs)
# print(steps_per_epoch)
training_loss = []
for st in range(epochs+1):
    if st == 0:
        training_loss.append(data[1][st])
    else:
        training_loss.append(data[1][(st*steps_per_epoch)-1])

total_epochs = [i for i in range(epochs+1)]
epochs_val = total_epochs.copy()
del epochs_val[0]
val_loss = data[-1]

# Ploting loss and accuracy graphs

In [ ]:

figure, axis = plt.subplots(2, 2)
# For Train Loss
axis[0, 0].plot(total_epochs, training_loss, label='Train Loss')
axis[0, 0].set_title("Train Loss")
# For Test Loss
axis[0, 1].plot(epochs_val, val_loss, label='Test Loss')
axis[0, 1].set_title("Test Loss")
# For Test Loss
axis[1, 0].plot(epochs_val, data_acc, label='Test Accuracy')
axis[1, 0].set_title("Test Accuracy")
# For Overall Model Evaluation
axis[1, 1].plot(total_epochs, training_loss, label='Train Loss')
axis[1, 1].plot(epochs_val, val_loss, label='Test Loss')
axis[1, 1].plot(epochs_val, data_acc, label='Test Accuracy')
axis[1, 1].set_title("All Training Plots")
axis[1, 1].legend()
plt.savefig('resnet152_fastai_plot.png')

In [ ]:
# calculate the accuracy
preds,y, loss = learn.get_preds(with_loss=True)
acc = accuracy(preds, y)
print('The accuracy is {0} %.'.format(acc))


# Save and Export The model

In [ ]:

learn.save('/content/drive/MyDrive/cars_classification/models_resnet_152/stanford-cars-1_resnet152')
learn.export('/content/drive/MyDrive/cars_classification/models_resnet_152/export.pkl')